
#' Testing Associations between Phenotype and Predicted Host Gene Expression
#'
#' @description MicroGenixAssociation performs association tests between phenotype and gene expression levels predicted using MicroGenixPredict.
#'
#' @param pred_expr a named numeric vector of predicted gene expression generated by MicroGenixPredict.
#' @param metadata the input file of metadata.
#' @param pheno_type type of phenotype. 'binary'(binary traits) or 'quant'(quantitative traits)
#' @param model the file name of a model object (.rds). Optional, if provided, interaction terms are returned.
#' @param pheno the column name of phenotype in the metadata file.
#' @param fixed_effect a character or character vector of the column names for the fixed effects (apart from pheno) in the metadata file.
#' @param random_effect a character of the column name for the random effects in the metadata file.
#' @param sig_thres significance threshold for the association. Interaction terms are not returned for insignificant associations.
#'
#' @return beta and p-value from the association test. Interaction terms for significant associations (if model is provided).
#' @export
#'
#' @importFrom glmnet coef.glmnet
#' @importFrom stats as.formula
#' @importFrom stats binomial
#' @importFrom stats glm
#' @importFrom stats lm
#' @importFrom stats predict
#' @importFrom utils read.csv
#' @importFrom utils write.table
#'
#' @examples
#' \dontrun{
#' input_taxa <- 'extdata/example_taxon_abundance.csv'
#' input_geno <- 'extdata/example_genotype_dosage.csv'
#' model <- 'example_output_model.rds'
#' predicted_data <- MicroGenixPredict(model, input_taxa, input_geno)
#' pred_expr <- predicted_data$predicted_expr
#' metadata <- 'extdata/example_metadata.csv'
#' assoc_result <- MicroGenixAssociation(pred_expr, metadata, pheno = 'pheno')
#' }
MicroGenixAssociation <- function(pred_expr, metadata, pheno_type = 'binary', model = NULL,
                                  pheno = NULL, fixed_effect = NULL, random_effect = NULL,
                                  sig_thres = 0.05){
  metadata <- utils::read.csv(metadata, check.names = F, row.names = 1)
  if(! pheno_type %in% c('binary', 'quant')){
    stop('pheno_type should be: binary/quant')
  }
  sample_names <- intersect(names(pred_expr), rownames(metadata))
  if(length(sample_names) == 0){
    stop("Samples in the metadata file do not match")
  }
  metadata <- metadata[sample_names,]
  pred_expr <- pred_expr[sample_names]
  metadata$expr <- pred_expr
  if(is.null(pheno)){
    stop("The column name for the phenotype should be specified")
  }
  fmla <- paste(pheno, "~ expr")
  if(!is.null(fixed_effect)){
    fmla <- paste(fmla, '+', paste(fixed_effect, collapse = '+'))
  }
  if(!is.null(random_effect)){
    fmla <- paste(fmla, '+', '(1 |', random_effect, ')')
  }
  fmla <- stats::as.formula(fmla)
  if(pheno_type == 'binary'){
    assoc <- summary(stats::glm(fmla, data = metadata, family = binomial))
    beta <- assoc$coefficients['expr', 'Estimate']
    pval <- assoc$coefficients['expr', 'Pr(>|z|)']
  } else {
    assoc <- summary(stats::lm(fmla, data = metadata))
    beta <- assoc$coefficients['expr', 'Estimate']
    pval <- assoc$coefficients['expr', 'Pr(>|t|)']
  }
  if(!is.null(model) & pval < sig_thres){
    fit <- readRDS(model)
    fit_coef <- as.matrix(glmnet::coef.glmnet(fit, s="lambda.min"))
    fit_coef <- fit_coef[fit_coef[,1] != 0,]
    fit_coef <- fit_coef[grep('__x__', names(fit_coef), fixed = T)]
  } else {
    fit_coef <- NULL
  }
  result <- list(beta = beta, pval = pval, interact_coef = fit_coef)
  return(result)
}
